presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3InputStream.java

```diff
@@ -0,0 +1,363 @@
+            return retry()
+                    .maxAttempts(maxAttempts)
+                    .exponentialBackoff(BACKOFF_MIN_SLEEP, maxBackoffTime, maxRetryTime, 2.0)
+                    .stopOn(InterruptedException.class, PrestoS3FileSystem.UnrecoverableS3OperationException.class, EOFException.class, FileNotFoundException.class, AbortedException.class)
+                    .onRetry(stats::newGetObjectRetry)
+                    .run("getS3Object", () -> getS3Object(position, buffer, offset, length));
+        }
+        catch (Exception e) {
+            throw propagate(e);
+        }
+    }
+
+    private int getS3Object(long position, byte[] buffer, int offset, int length)
+            throws IOException
+    {
+        InputStream stream;
+        try {
+            GetObjectRequest request = new GetObjectRequest(host, keyFromPath(path))
+                    .withRange(position, (position + length) - 1);
+            stream = s3.getObject(request).getObjectContent();
+        }
+        catch (AmazonS3Exception e) {
+            stats.newGetObjectError();
+            switch (e.getStatusCode()) {
+                case HTTP_RANGE_NOT_SATISFIABLE:
+                    throw new EOFException(CANNOT_SEEK_PAST_EOF);
+                case HTTP_NOT_FOUND:
+                    throw new FileNotFoundException("File does not exist: " + path);
+                case HTTP_FORBIDDEN:
+                case HTTP_BAD_REQUEST:
+                    throw new PrestoS3FileSystem.UnrecoverableS3OperationException(path, e);
+                default:
+                    throw e;
+            }
+        }
+        catch (RuntimeException e) {
+            stats.newGetObjectError();
+            throw e;
+        }
+
+        stats.connectionOpened();
+        try {
+            int read = 0;
+            while (read < length) {
+                int n = stream.read(buffer, offset + read, length - read);
+                if (n <= 0) {
+                    if (read > 0) {
+                        return read;
+                    }
+                    return -1;
+                }
+                read += n;
+            }
+            return read;
+        }
+        catch (Throwable t) {
+            stats.newReadError(t);
+            abortStream(stream);
+            throw t;
+        }
+        finally {
+            stats.connectionReleased();
+            stream.close();
+        }
+    }
+
+    @Override
+    public void seek(long pos)
+            throws IOException
+    {
+        checkClosed();
+        if (pos < 0) {
+            throw new EOFException(NEGATIVE_SEEK);
+        }
+
+        // this allows a seek beyond the end of the stream but the next read will fail
+        nextReadPosition = pos;
+    }
+
+    @Override
+    public long getPos()
+    {
+        return nextReadPosition;
+    }
+
+    @Override
+    public int read()
+    {
+        // This stream is wrapped with BufferedInputStream, so this method should never be called
```

Comment:
This is a public class so this is not guaranteed.

Reply:
This a direct refactor from the original class which was private. I've removed the comment